subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionReport.kt

```diff
@@ -68,33 +71,31 @@ class InstantExecutionReport(
 
         override fun buildFinished(result: BuildResult) {
             if (problems.isNotEmpty()) {
-                val outputDirectory = calculateOutputDirectory()
-                logSummary(outputDirectory)
-                writeReportFiles(outputDirectory)
+                if (result.failure?.isOrHasCause(InstantExecutionException::class) == true) {
```

Comment:
I would use a new kind of listener interface, so that the report receives all IE problems as events as the problems happen, rather than trying to unpack the exception chain and making assumptions about whether and how many context exception there are at the front of the chain. The report write can listen for events on this interface and then write the report at the end of the build.

Reply:
Makes sense, thanks. It'll simplify things further. The commented code changed with recent changes but this implementation issue is still there.

Reply:
I ended up untangling participants with the help of a listener. Things look much better now I think, thanks again.